### PR TITLE
Util Code review 2

### PR DESCRIPTION
**Pull request by jpatton-USGS**
* **Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?**
This PR contains the final round of changes and fixes to the util library based on the ongoing scientific software package review prior to release

* **Other information**:
This PR addresses the reviewer's pass of comments on the util library, which included: simplification, standardization, and removal of unused legacy functions.

The reviewer will review and approve that these changes address their comments.

### CODE REVIEW

**Comment by dkragness-usgs** [comment]
Woohoo!  Only 41 files changed.  And here I was afraid this wasn't gonna be easy...

**Comment by jpatton-USGS** [comment]
I can't submit a PR that breaks the build, sorry.

File: util/include/config.h [hunk 1]
```diff
@@ -87,7 +88,8 @@ class Config {
 	 * \throws For possible exceptions passed through this function see
 	 * Config::parseJSONFromString and Config::openFile
 	 */
-	json::Object parseJSONFromFile(std::string filepath, std::string filename);
+	std::shared_ptr<json::Object> parseJSONFromFile(std::string filepath,
```
Comment (dkragness-usgs) [inline util/include/config.h]: John and I have discussed the return values from these functions.  I believe they should be shared_ptr(const json::Object) and he likes them at least temporarily as shared_ptr(json::Object).
John is possibly still evaluating this, but we have agreed to keep them as shared_ptr(json::Object), until I get further along in the code review.

Comment (dkragness-usgs) [reply util/include/config.h]: My reasoning(before I forget again), is that an outside caller should not be able to change the underlying json object, without calling an update-esque method on the Config object.

File: util/include/threadbaseclass.h [hunk 2]
```diff
@@ -285,33 +259,16 @@ class ThreadBaseClass : public util::BaseClass {
...
-	 */
-	std::atomic<bool> m_bThreadHealth;
-
-	/**
-	 * \brief the std::mutex used to control access to m_bThreadHealth
+	 * \brief glass3::util::ThreadState enumeration used to track thread status,
+	 * set by setThreadState()
 	 */
-	std::mutex m_HealthCheckMutex;
+	std::atomic<glass3::util::ThreadState> m_bThreadState;
```
Comment (dkragness-usgs) [inline util/include/threadbaseclass.h]: should this really be named m_bThreadState?  doesn't seem like a boolean value.  
And yes... the irony of me complaining about a variable with a prefix of "b" having more than two values is not lost on me...

File: util/include/threadpool.h [hunk 3]
```diff
@@ -202,6 +196,17 @@ class ThreadPool : public util::BaseClass {
 	 */
 	const std::string& getPoolName();
 
+	/**
+	 * \brief Function to get thread states
+	 *
+	 * This function gets the thread states by getting the value of
```
Comment (dkragness-usgs) [inline util/include/threadpool.h]: instead of saying "gets the thread states", I think it would be less ambiguous if it said, "gets the state of the thread pool", since there's only one underlying state, whereas the other way, it sounds like there could be one per pool thread.

File: util/include/threadpool.h [hunk 4]
```diff
@@ -227,28 +232,28 @@ class ThreadPool : public util::BaseClass {
...
-	 * \brief Checks to see if the job threads in the pool should still be
-	 * running
+	 * \brief Function to retrieve the last time any of the job threads health
+	 * status was checked
 	 *
-	 * This function checks to see if the job thread should still running by
-	 * returning the value of m_bRunJobLoop.
-	 * \return Returns true if the job threads should still running, false if it
-	 * they should be stopped
+	 * This function retrieves the last time any of the the health statuses of
```
Comment (dkragness-usgs) [inline util/include/threadpool.h]: should be something like:
"retrieves the oldest time that any of the pool threads last updated their status as healthy."
as it's not reporting when status was checked, it's reporting when status was reported as healthy by each pool thread.
Status is checked by healthCheck(), and it doesn't update any values when it runs.

File: util/include/threadpool.h [hunk 5]
```diff
@@ -111,45 +116,25 @@ class ThreadPool : public util::BaseClass {
...
-	/**
-	 *\brief Retrieve the current number of jobs in the queue
-	 *
-	 * Retrieves the number of pending (queued but not started) jobs stored in
-	 * m_JobQueue
-	 *
-	 *\return an integer containing the current number of jobs in the queue
-	 */
-	int getJobQueueSize();
+	void setJobHealth(bool health = true);
```
Comment (dkragness-usgs) [inline util/include/threadpool.h]: I don't see any code that calls setJobHealth(false);
At the very least, it should be added to unittest for this class.

File: util/include/threadpool.h [hunk 6]
```diff
@@ -308,12 +292,7 @@ class ThreadPool : public util::BaseClass {
 	 * \brief A std::map containing the boolean status flags for each thread in
```
Comment (dkragness-usgs) [inline util/include/threadpool.h]: needs to be updated to reflect change of variable to be epoch times instead of bool flags.

File: util/include/threadstate.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * This enumeration defines the values used in tracking the state of a thread.
+ */
+enum ThreadState {
+	Initialized = 0, /**< thread has been initialized */
+	Starting = 1, /**< thread is starting up */
+	Started = 2, /**< thread has been started */
+	Stopping = -1, /**< thread is shutting down */
+	Stopped = -2	/**< thread is shut down */
+};
```
Comment (dkragness-usgs) [inline util/include/threadstate.h]: doesn't seem like you need it now, but might be worth considering having an "idle" state, if you wanted to monitor how busy threads were, or whether all work is done, etc.
Wouldn't work with the way status is done in threadpool, but just a thought.

File: util/src/cache.cpp [hunk 8]
```diff
@@ -48,33 +48,25 @@ bool Cache::addToCache(std::shared_ptr<json::Object> data, std::string id) {
 
 	// don't do anything if we didn't get an ID
 	if (id == "") {
-		logger::log("error", "cache::addstationtocache(): Bad id passed in.");
```
Comment (dkragness-usgs) [inline util/src/cache.cpp]: DK pickup here.

File: util/tests/baseclass_unittest.cpp [hunk 9]
```diff
@@ -17,10 +17,12 @@ TEST(BaseClassTest, CombinedTest) {
 
 	// generate configuration
 	std::string configstring = std::string(TESTCONFIG);
-	json::Object configuration = json::Deserialize(configstring);
+	json::Object* configuration = new json::Object(
```
Comment (dkragness-usgs) [inline util/tests/baseclass_unittest.cpp]: seems like this should be changed to:
std::shared_ptr<json::Object> sharedConfig(new json::Object(json::Deserialize(configstring))
instead of naming the intermediate pointer and giving someone a chance to delete it (which would mess things up).

File: util/tests/config_unittest.cpp [hunk 10]
```diff
@@ -10,23 +10,24 @@
 #define BADFILENAME "bad.d"
 #define BADCONFIGSTRING "bad"
 
-void checkdata(json::Object ConfigObject) {
+void checkdata(std::shared_ptr<json::Object> ConfigObject) {
```
Comment (dkragness-usgs) [inline util/tests/config_unittest.cpp]: seems like this should be defined as
void checkdata(std::shared_ptr<const json::Object> ConfigObject) {

unless you have plans to change the contents of ConfigObject in this function.

File: util/tests/config_unittest.cpp [hunk 11]
```diff
@@ -49,7 +50,7 @@ TEST(ConfigTest, TestFileLoading) {
 	glass3::util::Config * TestConfig2 = new glass3::util::Config();
 
 	// assert empty config object
-	ASSERT_TRUE(TestConfig2->getJSON().size() == 0)<< "empty config object";
+	ASSERT_TRUE(TestConfig2->getJSON() == NULL) << "empty config object";
```
Comment (dkragness-usgs) [inline util/tests/config_unittest.cpp]: this asks if the shared_ptr returned by getJSON() has a NULL pointer for it's internal pointer.  This is true when the shared_ptr is created with the default constructor(as would happen during config initialization).

File: util/tests/threadbaseclass_unittest.cpp [hunk 12]
```diff
@@ -141,16 +139,12 @@ TEST(ThreadBaseClassTest, CombinedTest) {
 	ASSERT_TRUE(TestThreadBaseStub->stop())<< "stop was successful";
 
 	// assert that class not started
-	ASSERT_FALSE(TestThreadBaseStub->isStarted())<<
-	"TestThreadBaseStub not started";
-
-	// assert that class not running
-	ASSERT_FALSE(TestThreadBaseStub->isRunning())<<
-	"TestThreadBaseStub not running";
+	ASSERT_TRUE(TestThreadBaseStub->getThreadState() ==
```
Comment (dkragness-usgs) [inline util/tests/threadbaseclass_unittest.cpp]: don't you have to wait a little while to execute this?
Couldn't it come back as "ThreadState::Stopping" sometimes if the threads don't start up quickly enough?

File: util/tests/threadbaseclass_unittest.cpp [hunk 13]
```diff
@@ -141,16 +139,12 @@ TEST(ThreadBaseClassTest, CombinedTest) {
...
+			glass3::util::ThreadState::Stopped)<< "TestThreadBaseStub not started";
 
 	// assert that healthCheck is true
```
Comment (dkragness-usgs) [inline util/tests/threadbaseclass_unittest.cpp]: Isn't this now asserting that the health check is false (has been stopped), instead of true?

File: util/tests/timeutil_unittest.cpp [hunk 14]
```diff
@@ -25,6 +35,24 @@ TEST(TimeUtil, ConvertEpochTimeToISO8601) {
 	ConvertedISO8601 = glass3::util::convertEpochTimeToISO8601(tEpochTime,
 																dDecimalSec);
 	ASSERT_STREQ(ConvertedISO8601.c_str(), ExpectedISO8601.c_str());
+
+	// edge case 1
```
Comment (dkragness-usgs) [inline util/tests/timeutil_unittest.cpp]: :+1:

File: util/tests/config_unittest.cpp (hunk 10 above)
Comment (jpatton-USGS) [reply util/tests/config_unittest.cpp]: I don't understand, when you create a shared_ptr, you always have to give it an inner type

File: util/tests/threadbaseclass_unittest.cpp (hunk 12 above)
Comment (jpatton-USGS) [reply util/tests/threadbaseclass_unittest.cpp]: k

File: util/tests/threadbaseclass_unittest.cpp (hunk 13 above)
Comment (jpatton-USGS) [reply util/tests/threadbaseclass_unittest.cpp]: yes

File: util/src/cache.cpp [hunk 15]
```diff
@@ -48,33 +48,25 @@ bool Cache::addToCache(std::shared_ptr<json::Object> data, std::string id) {
...
+		logger::log("error", "cache::addToCache(): Bad id passed in.");
 		return (false);
 	}
 
 	// lock in case someone else is using the cache
 	std::lock_guard<std::mutex> guard(getMutex());
 
+	// make a copy for the cache
+	std::shared_ptr<json::Object> newData = std::make_shared<json::Object>(
```
[the remark this answers is not in the document]
Comment (jpatton-USGS) [reply util/src/cache.cpp]: const what? shared_ptr? I'm not sure that means what you think it means, I think the const only applies to the shared_ptr, not the json object it points to

File: util/tests/config_unittest.cpp (hunk 10 above)
Comment (dkragness-usgs) [reply util/tests/config_unittest.cpp]: I think I meant to say something about it should be a const pointer, but somehow I hosed the comment.

Something like:
void checkdata(std::shared_ptr<const json::Object> ConfigObject)

File: util/src/timeutil.cpp [hunk 16]
```diff
@@ -29,15 +29,19 @@ std::string convertEpochTimeToISO8601(double epochtime) {
 	double integerpart;
 	double fractionpart;
 
+	// split the double
 	fractionpart = modf(epochtime, &integerpart);
 
-	time_t time = static_cast<int>(integerpart);
+	// modf returns only the integer part (3.14159265 -> 3.000000)
+	// so a simple cast is sufficient
+	std::time_t time = static_cast<int>(integerpart);
```
[the remark this answers is not in the document]
Comment (jpatton-USGS) [reply util/src/timeutil.cpp]: But I WANT it to truncate, that's the point of the modf()! The truncated stuff after the decimal point is saved and passed in fractionpart

File: util/src/cache.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (dkragness-usgs) [reply util/src/cache.cpp]: If you make the return parameter from getCache() or whatever the functions are called that return data from the cache via a public function : 
std::shared_ptr<const json::Object>

then I think you don't need to copy the objects when you export them.

File: util/tests/config_unittest.cpp (hunk 10 above)
Comment (jpatton-USGS) [reply util/tests/config_unittest.cpp]: you need to use code tags for this, otherwise the markdown interpreter eats them

File: util/src/timeutil.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (dkragness-usgs) [reply util/src/timeutil.cpp]: I think you're confused about what I'm saying.
What you're wanting to do is separate the integer portion of the seconds value from the decimal portion.
The problem is that modf() doesn't return any integers, it only returns floats(or probably doubles), and floating point values are imperfect representations of integers.
My concern is that once you've separated out the integerpart, you're gonna end up with something that is slightly less than the integer you want to represent with integerpart, and if you cast that to an integer, you may not get the value you are expecting, while if you call rint(), you will.
Take another look at the example in my comment.

File: util/src/threadpool.cpp [hunk 17]
```diff
@@ -192,110 +185,101 @@ void ThreadPool::jobLoop() {
...
+
+	// go through all threads in the pool
+	// I don't think we need a mutex here because the only function that
+	// can modify the size of a map is start()
+	std::map<std::thread::id, std::atomic<double>>::iterator StatusItr;
+	for (StatusItr = m_ThreadHealthMap.begin();
+			StatusItr != m_ThreadHealthMap.end(); ++StatusItr) {
 		// get the thread status
-		bool status = static_cast<bool>(StatusItr->second);
+		double healthTime = static_cast<double>(StatusItr->second);
```
[the remark this answers is not in the document]
Comment (jpatton-USGS) [reply util/src/threadpool.cpp]: Codacy seems to want me to reduce variable scope to it's minimum

File: util/src/threadbaseclass.cpp [hunk 18]
```diff
@@ -48,8 +48,8 @@ ThreadBaseClass::~ThreadBaseClass() {
 // ---------------------------------------------------------start
 bool ThreadBaseClass::start() {
 	// are we already running
-	if ((getThreadState() == glass3::util::ThreadState::Starting)
-			|| (getThreadState() == glass3::util::ThreadState::Started)) {
+	if ((getThreadState() != glass3::util::ThreadState::Initialized)
+			|| (getThreadState() != glass3::util::ThreadState::Stopped)) {
```
Comment (dkragness-usgs) [inline util/src/threadbaseclass.cpp]: This should be an AND (&&) instead of an OR(||)  since it's a != comparison, otherwise everything will pass.

File: util/src/threadpool.cpp [hunk 19]
```diff
@@ -38,8 +38,8 @@ ThreadPool::~ThreadPool() {
 // ---------------------------------------------------------start
 bool ThreadPool::start() {
 	// are we already running
-	if ((getThreadPoolState() == glass3::util::ThreadState::Starting)
-			|| (getThreadPoolState() == glass3::util::ThreadState::Started)) {
+	if ((getThreadPoolState() != glass3::util::ThreadState::Initialized)
+			|| (getThreadPoolState() != glass3::util::ThreadState::Stopped)) {
```
Comment (dkragness-usgs) [inline util/src/threadpool.cpp]: && not ||

**Comment by jpatton-USGS** [comment]
I think you reviewed early, or didn't refresh, your change comments were immediatly marked as outdated.

File: util/src/timeutil.cpp [hunk 20]
```diff
@@ -34,7 +34,8 @@ std::string convertEpochTimeToISO8601(double epochTime) {
 
 	// modf returns only the integer part (3.14159265 -> 3.000000)
 	// so a simple cast is sufficient
-	std::time_t time = static_cast<int>(integerpart);
+	std::time_t time = static_cast<int>(
+			integerpart >= 0.0 ? (integerpart + 0.1) : (integerpart - 0.1));
```
Comment (dkragness-usgs) [inline util/src/timeutil.cpp]: really?  rint() was just a bridge too far?  :3rd_place_medal:   :smiley_cat:
